doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,152 @@
+# Read Holds
+
+- Durable Subscribes epic: [17541](https://github.com/MaterializeInc/materialize/issues/17541)
+- Configurable compaction windows: [16701](https://github.com/MaterializeInc/materialize/issues/16701)
+
+## The Problem
+
+Users are not able to use `SUBSCRIBE` without data loss.
+When materialize restarts or a client connection fails for some reason,
+users are not able to restart the `SUBSCRIBE` `AS OF` their most recently observed timestamp because the read frontier has usually progressed past that time.
+This happens because the compaction window is 1 second for all user objects.
+
+## Success Criteria
+
+`SUBSCRIBE` must be usable without data loss across server restarts or client reconnections.
+
+## Solution Proposal
+
+Add a new `HOLD` object that acquires controller read holds on listed objects at a specific timestamp.
+The `HOLD` must be manually advanced to newer timestamps by users which will release the previous and acquire a new read hold to the specified timestamp.
+If the timestamp falls behind a configurable window, the `HOLD` is marked in an error state and the read holds are released.
+
+The `HOLD` can be created by:
+
+```
+CREATE HOLD <name>
+ON <object1>[, <object2> ...]
+[WITH (
+    AT TIME = <time>,
```

Comment:
```suggestion
    AT = <time>,
```

Using `AT = <time>` aligns a bit better with the forthcoming `REFRESH AT <time>` syntax.

There's also the argument for using the special syntax you proposed initially (`CREATE HOLD <name> ON <object> AT <time>`), for what it's worth. Something like `AT` being such a fundamental part of holds that it deserves its own syntax. That justifies having special advancement syntax too (`ALTER HOLD ... ADVANCE TO <time>`).

Now that I see it all written out, I think we either want `CREATE and `ALTER` to both be standard...

```sql
CREATE HOLD <name> ON <object> WITH (AT = <time>);
ALTER HOLD <name> SET (AT = <time>);
```

Or all special:

```sql
CREATE HOLD <name> ON <object> AT <time>;
ALTER HOLD <name> ADVANCE TO <time>;
```

Sorry for inducing you to go and change it. In my original comment, I was genuinely asking whether `WITH (AT ...)` would be better, not expressing an opinion! 🙈